### PR TITLE
go-selinux: add coverage for LfileLabel, LsetFileLabel

### PR DESCRIPTION
**Pull request by thaJeztah**
updates: 1b18907dde2de75c9762d425c862586e1f7c438a (https://github.com/opencontainers/selinux/pull/169)

### CODE REVIEW

**Comment by rhatdan** [comment]
LGTM

File: go-selinux/selinux_linux_test.go [hunk 1]
```diff
@@ -16,25 +16,61 @@ func TestSetFileLabel(t *testing.T) {
...
+	if filelabel != con {
 		t.Fatalf("FileLabel failed, returned %s expected %s", filelabel, con)
 	}
+
+	// Using LfileLabel to verify that the symlink itself is not labeled.
+	linkLabel, err := LfileLabel(tmpLink)
+	if err != nil {
+		t.Fatalf("LfileLabel failed: %s", err)
+	}
+	if linkLabel != "" {
```
Comment (rhatdan) [inline go-selinux/selinux_linux_test.go:51]: This should be linkLabel == con

File: go-selinux/selinux_stub_test.go [hunk 2]
```diff
@@ -22,6 +22,13 @@ func TestSELinuxStubs(t *testing.T) {
 		t.Error(err)
 	}
 
+	if _, err := LfileLabel("/etc"); err != nil {
+		t.Error(err)
+	}
+	if err := LsetFileLabel("/etc", testLabel); err != nil {
```
Comment (rhatdan) [inline go-selinux/selinux_stub_test.go:28]: This could be dangerous if run as root.

Comment (thaJeztah) [reply go-selinux/selinux_stub_test.go:28]: Yes, I was looking for a good path to use (this is for the stubs, so _should_ not do anything, but the existing tests also used `/etc`)

File: go-selinux/selinux_linux_test.go (hunk 1 above)
Comment (thaJeztah) [reply go-selinux/selinux_linux_test.go:51]: My thinking here was that no label should be set _at all_ on the symlink, correct?

**Comment by thaJeztah** [comment]
(oh! forgot to post); Hmm... right, so this won't help a lot; while looking, of course I stumbled upon this ticket from @kolyshkin 😂  https://github.com/actions/runner-images/issues/2307

Also had a look if we could install SELinux on Ubuntu, but it requires a reboot, so not an option 😞 

```
=== RUN   TestSetFileLabel
    selinux_linux_test.go:16: SELinux not enabled, skipping.
--- SKIP: TestSetFileLabel (0.00s)
=== RUN   TestKVMLabels
    selinux_linux_test.go:78: SELinux not enabled, skipping.
--- SKIP: TestKVMLabels (0.00s)
=== RUN   TestInitLabels
    selinux_linux_test.go:99: SELinux not enabled, skipping.
--- SKIP: TestInitLabels (0.00s)
=== RUN   TestSELinux
    selinux_linux_test.go:131: SELinux not enabled, skipping.
--- SKIP: TestSELinux (0.00s)
...
```

File: go-selinux/selinux_stub_test.go (hunk 2 above)
Comment (thaJeztah) [reply go-selinux/selinux_stub_test.go]: Updated to use a temp-file to test with 👍

File: go-selinux/selinux_linux_test.go (hunk 1 above)
Comment (rhatdan) [reply go-selinux/selinux_linux_test.go:51]: All files have labels.  Getting a file without a label on an SELinux system is difficult.

File: go-selinux/selinux_stub_test.go (hunk 2 above)
Comment (rhatdan) [reply go-selinux/selinux_stub_test.go]: If we are using system directories elsewhere we probably should not.

**Comment by rhatdan** [comment]
We really should be testing on Centos or Fedora.

**Comment by rhatdan** [comment]
Apply this patch to yours and the tests passes.
[diff.patch](https://github.com/opencontainers/selinux/files/9923538/diff.patch.txt)

**Comment by kolyshkin** [comment]
> Also had a look if we could install SELinux on Ubuntu, but it requires a reboot, so not an option

I remember @avagin implemented a complicated trick with upgrading the kernel and rebooting a travis machine while keeping the process that connects to travis infra intact (using CRIU), which helped us to test newer kernels.

Today, we can use something like cirrus-ci which you can use to install centos (7, 8, or 9) and also run fedora via KVM. See runc's .cirrus.yml for example.

We can also use github ci to run fedora in a KVM, but this is more error prone since the only way is to use Mac OS X and there's a 5 to 10% chance that it will fail, based on what I saw.

File: go-selinux/label/label_stub_test.go [hunk 3]
```diff
@@ -75,15 +79,22 @@ func TestProcessLabel(t *testing.T) {
 }
 
 func TestCheckLabelCompile(t *testing.T) {
+	tmpFile := filepath.Join(t.TempDir(), "test_file")
+	out, err := os.OpenFile(tmpFile, os.O_WRONLY|os.O_CREATE, 0)
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (kolyshkin) [inline go-selinux/label/label_stub_test.go:86]: Why not merely use `t.TempDir()`. It returns a directory name which you can use directly as an argument to `FileLabel` and `SetFileLabel`.

File: go-selinux/selinux_stub_test.go [hunk 4]
```diff
@@ -13,12 +15,18 @@ func TestSELinuxStubs(t *testing.T) {
 	if GetEnabled() {
 		t.Error("SELinux enabled on non-linux.")
 	}
+	tmpFile := filepath.Join(t.TempDir(), "test_file")
```
Comment (kolyshkin) [inline go-selinux/selinux_stub_test.go:18]: ditto

**Review by kolyshkin — CHANGES_REQUESTED** [review]
A nits to simplify code by using t.TempDir more directly

**Comment by rhatdan** [comment]
@thaJeztah Still working on this?

**Comment by thaJeztah** [comment]
Oh! Looks like I somehow missed the comments here; updated 👍

File: go-selinux/label/label_stub_test.go [hunk 5]
```diff
@@ -3,7 +3,9 @@
 
 package label
 
-import "testing"
+import (
+	"testing"
+)
```
Comment (thaJeztah) [inline go-selinux/label/label_stub_test.go:8]: Oh! Let me revert this part

Comment (thaJeztah) [reply go-selinux/label/label_stub_test.go]: done 👍

**Comment by rhatdan** [comment]
LGTM
@kolyshkin @giuseppe @vrothberg PTAL

**Comment by rhatdan** [comment]
@lsm5 PTAL

**Review by rhatdan — APPROVED** [review]

**Comment by thaJeztah** [comment]
> @lsm5 PTAL

Oh! Voice from the past. It's been a while since we interacted (hope you're doing well!)

**Comment by rhatdan** [comment]
@lsm5 has shown interest in SELinux so I figured I would ping him.

**Comment by lsm5** [comment]
> > @lsm5 PTAL
> 
> Oh! Voice from the past. It's been a while since we interacted (hope you're doing well!)

@thaJeztah :wave: I'm good, hope you are too! :smile:

> @lsm5 has shown interest in SELinux so I figured I would ping him.

@rhatdan shown interest yes, but this is outta my reviewing capacity atm, LGTM from a cursory look though. Also `sudo make test` passes for this PR on my F37 env.

Also, RE: fedora/centos testing, do you want that to be a CI task here?

**Comment by rhatdan** [comment]
It would be awesome if we could test Fedora or Centos here.

**Comment by lsm5** [comment]
@rhatdan IIUC, github actions is ubuntu only. So, we'll likely need some other setup. 

@cevich wdyt? will it be a lot of pain to enable cirrus outside the containers org?

**Comment by thaJeztah** [comment]
Yeah, it's complicated. It's possible to attach custom runners to GitHub actions, but those would need to be maintained somewhere (_possibly_ the CNCF has infra for this?). Some caution is needed for these as well to prevent those runners from also be running on forks (and to be sure they're only running on "good" pull requests, to prevent crypto miners from using them).

Of course the ideal would be for GitHub to provide at least "one" variant of an rpm/yum/dnf flavoured distro.

**Pull request merged**